File: Java + Native + Spring Boot 3/src/main/java/ac/lsys/apples/Controller.java
package ac.lsys.apples;

import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.UUID;

@RestController
@RequestMapping("/{uuid}")
public class Controller {

    private final EventRepository eventRepository;

    public Controller(EventRepository eventRepository) {
        this.eventRepository = eventRepository;
    }

    @PostMapping("/command")
    Event handleCommand(@PathVariable("uuid") UUID uuid, @RequestBody Command command) {
        // Query all existing events
        List<Event> events = eventRepository.findByAggregateUuid(uuid);

        // Apply all existing events to the aggregate
        Aggregate aggregate = new Aggregate(uuid);
        events.forEach(aggregate::apply);

        // Create a new event from the command
        Event event = new Event(uuid);
        event.setData(command.getData());

        // Apply the new event to the aggregate
        aggregate.apply(event);

        // Store and return the new event
        return eventRepository.save(event);
    }

    @GetMapping("/query")
    String handleQuery(@PathVariable("uuid") UUID uuid) {
        // Query all existing events
        List<Event> events = eventRepository.findByAggregateUuid(uuid);

        // Apply all events to the projection
        Projection projection = new Projection();
        events.forEach(projection::apply);

        // Return the projection's state
        return projection.getState();
    }
}

File: Kotlin + JVM + Micronaut/src/main/kotlin/ac/lsys/application/Controller.kt
package ac.lsys.application

import ac.lsys.plugins.Aggregate
import ac.lsys.plugins.Command
import ac.lsys.plugins.Event
import ac.lsys.plugins.Projection
import io.micronaut.context.BeanContext
import io.micronaut.http.HttpResponse
import io.micronaut.http.MutableHttpResponse
import io.micronaut.http.annotation.Body
import io.micronaut.http.annotation.Controller
import io.micronaut.http.annotation.Get
import io.micronaut.http.annotation.PathVariable
import io.micronaut.http.annotation.Post
import io.micronaut.transaction.TransactionDefinition
import io.micronaut.transaction.annotation.Transactional
import java.util.UUID

@Controller
class Controller(private val beanContext: BeanContext) {


    @Post("/{uuid}/command")
    @Transactional(isolation = TransactionDefinition.Isolation.SERIALIZABLE)
    fun post(@PathVariable uuid: String, @Body command: Command): HttpResponse<Event> {

        val eventService = beanContext.getBean(EventRepository::class.java)

        val events: List<Event> = eventService.readByUuid(uuid)

        val aggregate = Aggregate(uuid)
        events.forEach(aggregate::apply)

        val newEvent = Event(uuid, command.data)
        aggregate.apply(newEvent)

        val eventId = eventService.create(newEvent)

        return HttpResponse.ok(eventId)
    }

    @Get("/{uuid}/query")
    fun query(@PathVariable uuid: String): HttpResponse<String>? {
        val eventService = beanContext.getBean(EventRepository::class.java)

        val events = eventService.readByUuid(uuid)

        val projection = Projection()
        events.forEach(projection::apply)

        return HttpResponse.ok<String>(projection.getState())
    }
}
